pkg/istio-agent/xds_proxy.go

```diff
@@ -504,6 +505,11 @@ func (p *XdsProxy) getTLSDialOption(agent *Agent) (grpc.DialOption, error) {
 	// strip the port from the address
 	parts := strings.Split(agent.proxyConfig.DiscoveryAddress, ":")
 	config.ServerName = parts[0]
+	// For debugging on localhost (with port forward)
+	// This matches the logic for the CA; this code should eventually be shared
+	if strings.Contains(config.ServerName, "localhost") {
+		config.ServerName = "istiod.istio-system.svc"
```

Comment:
I am not sure why the server name is  "istiod.istio-system.svc" not "localhost" when it is on localhost.

Reply:
IMO we should cleanup all these kinds of hack

Reply:
@hzxuzhonghu if you are running over localhost (local developement, or proxy running in same pod as istiod) the cert is signed for `istiod`, but we request localhost. This is a hack, sure, but we cannot remove it without a viable way to run over localhost